stdlib/public/core/ThreadLocalStorage.swift

```diff
@@ -0,0 +1,152 @@
+import SwiftShims
+
+// For testing purposes, a thread-safe counter to guarantee that destructors get
+// called by pthread.
+#if INTERNAL_CHECKS_ENABLED
+public // @testable
+let _destroyTLSCounter = _stdlib_AtomicInt()
+#endif
+
+// Thread local storage for all of the Swift standard library
+//
+// @moveonly/@pointeronly: shouldn't be used as a value, only through its
+// pointer. Similarly, shouldn't be created, except by
+// _initializeThreadLocalStorage.
+//
+internal struct _ThreadLocalStorage {
+  // TODO: might be best to absract uBreakIterator handling and caching into
+  // separate struct. That would also make it easier to maintain multiple ones
+  // and other TLS entries side-by-side.
+
+  // Save a pre-allocated UBreakIterator, as they are very expensive to set up.
+  // Each thread can reuse their unique break iterator, being careful to reset
+  // the text when it has changed (see below). Even with a naive always-reset
+  // policy, grapheme breaking is 30x faster when using a pre-allocated
+  // UBreakIterator than recreating one.
+  //
+  // private
+  var uBreakIterator: OpaquePointer
+
+  // TODO: Consider saving two, e.g. for character-by-character comparison
+
+  // The below cache key tries to avoid resetting uBreakIterator's text when
+  // operating on the same String as before. Avoiding the reset gives a 50%
+  // speedup on grapheme breaking.
+  //
+  // As a invalidation check, save the base address from the last used
+  // StringCore. We can skip resetting the uBreakIterator's text when operating
+  // on a given StringCore when both of these associated references/pointers are
+  // equal to the StringCore's. Note that the owner is weak, to force it to
+  // compare unequal if a new StringCore happens to be created in the same
+  // memory.
+  //
+  // private
+  weak var associatedStringOwner: AnyObject? = nil
+  // private
+  var associatedBaseAddress: UnsafeMutableRawPointer? = nil
+  // private
+  var associatedCountAndFlags: UInt = 0
+
+  // private: Should only be called by _initializeThreadLocalStorage
+  init(_uBreakIterator: OpaquePointer) {
+    self.uBreakIterator = _uBreakIterator
+  }
+
+  // Get the current thread's TLS pointer. On first call for a given thread,
+  // creates and initializes a new one.
+  static internal func getPointer()
+    -> UnsafeMutablePointer<_ThreadLocalStorage>
+  {
+    let tlsRawPtr = _swift_stdlib_pthread_getspecific(_tlsKey)
+    if _fastPath(tlsRawPtr != nil) {
+      return tlsRawPtr._unsafelyUnwrappedUnchecked.assumingMemoryBound(
+        to: _ThreadLocalStorage.self)
+    }
+
+    return _initializeThreadLocalStorage()
+  }
```

Comment:
I'm sure you already know... but if you move getPointer into the C runtime you can use clang's TLS support when available. That will remove all of the pthread call overhead.